clang/lib/Sema/SemaChecking.cpp

```diff
@@ -16548,11 +16548,10 @@ static void CheckImplicitConversion(Sema &S, Expr *E, QualType T,
         std::string PrettySourceValue = toString(Value, 10);
         std::string PrettyTargetValue = PrettyPrintInRange(Value, TargetRange);
 
-        S.DiagRuntimeBehavior(
-            E->getExprLoc(), E,
-            S.PDiag(diag::warn_impcast_integer_precision_constant)
-                << PrettySourceValue << PrettyTargetValue << E->getType() << T
-                << E->getSourceRange() << SourceRange(CC));
```

Comment:
This seems unrelated - We do not want to emit this diag during constant evaluation, because we know the source value and that it does not lose precision.

Reply:
I'm fine either way, but the following three warning diagnostics are lost without this change:
- [CXX/expr/expr.const/p2-0x.cpp#L247](https://github.com/llvm/llvm-project/blob/main/clang/test/CXX/expr/expr.const/p2-0x.cpp#L247)
- [CXX/expr/expr.const/p2-0x.cpp$248](https://github.com/llvm/llvm-project/blob/main/clang/test/CXX/expr/expr.const/p2-0x.cpp#L248)
- [AST/Interp/intap.cpp#L12](https://github.com/llvm/llvm-project/blob/main/clang/test/AST/Interp/intap.cpp#L12)

I think @erichkeane previously objected to losing these diagnostics.

Reply:
So, this is interesting.
The standard does not require the conversion to be ill-formed, so your change is actually correct.

Reply:
Its a warning, of course the standard doesn't require it.  It seems to me that it is a USEFUL diagnostic though.

Reply:
@cor3ntin Do you think any changes to diagnostics need to happen in this PR, or are you okay with the warning applying during constant evaluation?

Reply:
I'm fine with things as-is.
As a follow up we could: 
 - Make it a warning that default to an error
 - Try to remove the duplication in C and in the `constexpr char c = {1000}` case